src/Tizen.Diagnostics/Tizen.Diagnostics/Diagnostics.cs

```diff
@@ -0,0 +1,294 @@
+    /// </summary>
+    /// <since_tizen> 8 </since_tizen>
+    public class DiagnosticsContext : IDisposable
+    {
+        private IntPtr _handle = IntPtr.Zero;
+        private bool _disposed = false;
+
+        /// <summary>
+        /// The DiagnosticsContext class constructor.
+        /// </summary>
+        /// <since_tizen> 8 </since_tizen>
+        /// <param name="handle">Diagnostics context received in the ReportReceived event.</param>
+        internal DiagnosticsContext(IntPtr handle)
+        {
+            _handle = handle;
+
+            var ret = NativeApi.GetType(handle, out ContextType contextType);
+            if (ret != NativeApi.DiagnosticsError.None)
+                throw ExceptionFactory.CreateException(ret);
+
+            ContextType = contextType;
+        }
+
+        /// <summary>
+        /// Finalizes an instance of the DiagnosticsContext class.
+        /// </summary>
+        ~DiagnosticsContext()
+        {
+            Dispose(false);
+        }
+
+        /// <summary>
+        /// Gets report of specific type from the context.
+        /// </summary>
+        /// <since_tizen> 8 </since_tizen>
+        /// <param name="report">Report type to retrieve from the context.</param>
+        /// <returns>
+        /// The instance of DiagnosticsData.
+        /// </returns>
+        /// <remarks>
+        /// This function is permitted only to an app signed by platform level certificates.
+        /// </remarks>
+        public DiagnosticsData GetReport(ReportType report)
+        {
+            var ret = NativeApi.GetReport(_handle, report, out IntPtr data);
+            if (ret != NativeApi.DiagnosticsError.None)
+                throw ExceptionFactory.CreateException(ret);
+
+            return new DiagnosticsData(data);
+        }
+
+        /// <summary>
+        /// Stores context type.
+        /// </summary>
+        /// <since_tizen> 8 </since_tizen>
+        public ContextType ContextType { get; internal set; }
+
+        /// <summary>
+        /// Dispose API for closing the internal resources.
+        /// </summary>
+        /// <since_tizen> 8 </since_tizen>
+        public void Dispose()
+        {
+            Dispose(true);
+            GC.SuppressFinalize(this);
+        }
+
+        /// <summary>
+        /// Dispose API for closing the internal resources.
+        /// </summary>
+        /// <since_tizen> 8 </since_tizen>
+        protected virtual void Dispose(bool disposing)
+        {
+            if (!_disposed)
+            {
+                if (_handle != IntPtr.Zero)
+                {
+                    NativeApi.Destroy(_handle);
+                    _handle = IntPtr.Zero;
+                }
+                _disposed = true;
+            }
+        }
+    }
+
+    /// <summary>
+    /// The DiagnosticsData class is used for reading data (either from report or client).
+    /// </summary>
+    /// <since_tizen> 8 </since_tizen>
+    public class DiagnosticsData : IDisposable
```

Comment:
If this acts like a ReadOnly Stream, I suggest implementing System.IO.Stream.
Many open sources already implement Stream with ReadOnly, and the implementation can come out almost the same as this class just by throwing a NotSupport exception in the Write method.
Also, by implementing Stream, you can use many classes for Stream.

Reply:
> Many open sources already implement Stream with ReadOnly
Can you show some examples?